### PR TITLE
Add remove group

### LINKED ISSUE #119: Admin Forms Page
As an admin, I want to manage groups and users of the CCM website, so that they can perform their work.

[Story Points: 2]
[Priority: Medium]
[Risk: Medium]

This is a subtask of #21 

Demo steps:
Given I'm logged in as an admin,
From the home page I can view the admin section,
From the admin section I can go to a page to edit and manage groups,

### LINKED ISSUE #160: Add User to Group
As an admin, I want to add a user to a group.

[Story Points: 2]
[Priority: Medium]
[Risk: Medium]

Demo steps:
Given I'm logged in as an admin,
I can view the manage groups page,
then reach the add user page for a specific group,
and add a user that isn't already in that group.

### PR DESCRIPTION
**Pull request by Dinferno-lol**
Adds the add user to group page and requires admin rights to access it and the manage group page, functionality for group masters will be added on a separate page with future request with remove user.

closes #119

### CODE REVIEW

File: App/src/pages/ManageableGroups.tsx [hunk 1]
```diff
@@ -65,16 +65,21 @@ export default function DisplayManageableGroups() {
                                         {group.members?.length}
                                     </Td>
                                     <Td whiteSpace="nowrap" padding="16px">
-                                        <Link to={BaseRoutes.Home} state={{ gid: group.id, name: group.name }}>
+                                        <Link
+                                            to={BaseRoutes.AddUserToGroup}
```
Comment (JamesPartsafas) [inline App/src/pages/ManageableGroups.tsx:69]: There seems to be some bug with accessing the addUser page. If you click to go to addUsers for one group, then go back and click on another, the screen will often just turn white and you have to reload. This bug seems to happen fairly often, but not quite entirely reproducibly, just click back and forth a bit and you'll get it sooner or later.

File: App/src/pages/ManageableGroups.tsx [hunk 2]
```diff
@@ -65,16 +65,21 @@ export default function DisplayManageableGroups() {
...
                                                 Add
                                             </Button>
                                         </Link>
-                                        <Link to={BaseRoutes.Home} state={{ gid: group.id, name: group.name }}>
-                                            <Button style="primary" variant="outline" width="50%" height="40px">
-                                                Remove
-                                            </Button>
-                                        </Link>
+                                        {group.members.length != 0 && (
+                                            <Link to={BaseRoutes.Home} state={{ gid: group.id, name: group.name }}>
```
Comment (JamesPartsafas) [inline App/src/pages/ManageableGroups.tsx:77]: Is the removeUsers page done? If so, you should link to it.

File: App/src/App.tsx [hunk 3]
```diff
@@ -62,9 +69,34 @@ export function App() {
                     <Route path={BaseRoutes.Register} element={<Register setUser={setUser} />} />
                     <Route
                         path={BaseRoutes.ManageableGroup}
```
Comment (JamesPartsafas) [inline App/src/App.tsx:71]: In addition to having this route, can you also add a button to the home screen to bring the users to the manageable groups page? (As with the routes, keep it hidden unless the user is an admin or group master).

File: App/src/pages/addUserToGroup.tsx [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+                                        <li>{item.email}</li>
+                                        <Button
+                                            style="primary"
+                                            variant="outline"
+                                            width="22%"
+                                            height="40px"
+                                            justifyContent={"flex-end"}
+                                            onClick={() => addingUser(item.id)}
+                                        >
+                                            Select
```
Comment (JamesPartsafas) [inline App/src/pages/addUserToGroup.tsx:123]: Can you change the button content to "Add"?

File: App/src/pages/RemoveUserFromGroup.tsx [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+                                        <li>{item.email}</li>
+                                        <Button
+                                            style="primary"
+                                            variant="outline"
+                                            width="22%"
+                                            height="40px"
+                                            justifyContent={"flex-end"}
+                                            onClick={() => removingUser(item.id)}
+                                        >
+                                            Select
```
Comment (JamesPartsafas) [inline App/src/pages/RemoveUserFromGroup.tsx:103]: Can you change the button content to "Remove"?

File: App/src/pages/ManageableGroups.tsx (hunk 1 above)
Comment (Dinferno-lol) [reply App/src/pages/ManageableGroups.tsx:69]: I saw the bug, no idea how to fix it yet, unsure what's causing it

File: App/src/pages/ManageableGroups.tsx (hunk 2 above)
Comment (Dinferno-lol) [reply App/src/pages/ManageableGroups.tsx:77]: I wanted to make it a separate pull request, it is working but It's not displaying the users on load or removing them from the list automatically, I'm only able to have it update once the user starts typing in the search bar

File: App/src/pages/addUserToGroup.tsx (hunk 4 above)
Comment (Dinferno-lol) [reply App/src/pages/addUserToGroup.tsx:123]: sure

File: App/src/pages/RemoveUserFromGroup.tsx (hunk 5 above)
Comment (Dinferno-lol) [reply App/src/pages/RemoveUserFromGroup.tsx:103]: done

File: App/src/App.tsx (hunk 3 above)
Comment (Dinferno-lol) [reply App/src/App.tsx:71]: sure, was gonna add it with the next pull request with remove from group but I can add it here

File: App/src/App.tsx [hunk 6]
```diff
@@ -62,9 +69,34 @@ export function App() {
...
-                        element={isLoggedIn == true ? <DisplayManageableGroups /> : <Navigate to={BaseRoutes.Login} />}
+                        element={
+                            isLoggedIn == true && isAdmin == true ? (
```
[the remark this answers is not in the document]
Comment (Dinferno-lol) [reply App/src/App.tsx:73]: I renamed it for now, I hadn't gotten a chance to look at any of the other code implementing the jwt so I wasn't sure how to call it. I did get rid of the redundancy though so we can implement soon

File: App/src/pages/addUserToGroup.tsx [hunk 7]
```diff
@@ -0,0 +1,157 @@
+import { useEffect, useState } from "react";
```
[the remark this answers is not in the document]
Comment (Dinferno-lol) [reply App/src/pages/addUserToGroup.tsx:1]: I think I got a solution working with concat so it should add the other users to the end of it

File: App/src/App.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (Dinferno-lol) [reply App/src/App.tsx:73]: Ok I applied the jwt like you said, let me know if there are issues.

File: App/src/App.tsx [hunk 8]
```diff
@@ -19,13 +19,18 @@ import axios from "axios";
...
+import RemovingUserFromGroup from "./pages/RemoveUserFromGroup";
 import DeleteCourse from "./pages/DeleteCourse";
 
 export const UserContext = createContext<User | null>(null);
 
 export function App() {
     const [user, setUser] = useState<User | null>(initializeUser());
     const [isLoggedIn, setIsLoggedIn] = useState<boolean>(user != null ? true : false);
+    const [isAdminorGroupMaster, setIsAdminorGroupMaster] = useState<boolean>(
+        user != null ? (user.roles.includes("Admin") || user.masteredGroups != null ? true : false) : false
```
Comment (JamesPartsafas) [inline App/src/App.tsx:32]: user.masteredGroup should never be null. You should check if it's empty or not. If it's not empty, it means the user is a group master.

**Review by JamesPartsafas — APPROVED** [review]
LGTM

**Pull request merged**